src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -3466,15 +3466,27 @@ void MacroAssembler::cmpxchg(Register addr, Register expected,
   assert_different_registers(expected, t0);
   assert_different_registers(new_val, t0);
 
+  // NOTE:
+  // Register _result_ may be the same register as _new_val_ or _expected_.
+  // Hence do NOT use _result_ until after 'cas'.
+  //
+  // Register _expected_ may be the same register as _new_val_ and is assumed to be preserved.
+  // Hence do NOT change _expected_ or _new_val_.
+  //
+  // Having _expected_ and _new_val_ being the same register is a very puzzling cas.
+  //
+  // TODO: Address these issues.
+
   if (UseZacas) {
     if (result_as_bool) {
       mv(t0, expected);
       atomic_cas(t0, new_val, addr, size, acquire, release);
       xorr(t0, t0, expected);
       seqz(result, t0);
     } else {
-      mv(result, expected);
-      atomic_cas(result, new_val, addr, size, acquire, release);
+      mv(t0, expected);
+      atomic_cas(t0, new_val, addr, size, acquire, release);
```

Comment:
Not sure if I understand this correctly.
Could it be better to tell if `result` == `new_val`? If false, then we don't need the extra `mv`.

Reply:
The "mv()" check that them selfs, so an e.g. mv(a0, a0) will never emit an instruction.

OT NOTE:
If you are in an InCompressableRegion this will still not emit an instruction so your offsets may be wrong.

Reply:
Oh, sorry I now see what you mean.

Reply:
I see, let's do this optimization later after the fixing of these bugs.

Reply:
Great, thanks!